arraycontext/impl/pytato.py

```diff
@@ -36,11 +36,57 @@
 from numbers import Number
 import loopy as lp
 
+from arraycontext.container import serialize_container, is_array_container
 
-class _PytatoFakeNumpyLinalgNamespace(BaseFakeNumpyLinalgNamespace):
-    def norm(self, array, ord=None):
-        raise NotImplementedError
 
+class _PytatoFakeNumpyLinalgNamespace(BaseFakeNumpyLinalgNamespace):
+    def norm(self, ary, ord=None):
+        from numbers import Number
+        if isinstance(ary, Number):
+            return abs(ary)
+
+        if ord is None:
+            ord = 2
+
+        try:
+            from meshmode.dof_array import DOFArray
+        except ImportError:
+            pass
+        else:
+            if isinstance(ary, DOFArray):
+                from warnings import warn
+                warn("Taking an actx.np.linalg.norm of a DOFArray is deprecated. "
+                        "(DOFArrays use 2D arrays internally, and "
+                        "actx.np.linalg.norm should compute matrix norms of those.) "
+                        "This will stop working in 2022. "
+                        "Use meshmode.dof_array.flat_norm instead.",
+                        DeprecationWarning, stacklevel=2)
+
+                import numpy.linalg as la
+                return la.norm(
+                        [self.norm(_flatten_array(subary), ord=ord)
+                            for _, subary in serialize_container(ary)],
+                        ord=ord)
+
+        if is_array_container(ary):
+            import numpy.linalg as la
+            return la.norm(
+                    [self.norm(subary, ord=ord)
+                        for _, subary in serialize_container(ary)],
+                    ord=ord)
+
+        if len(ary.shape) != 1:
+            raise NotImplementedError("only vector norms are implemented")
+
+        if ary.size == 0:
+            return 0
+
+        if ord == np.inf:
+            return self._array_context.np.max(abs(ary))
+        elif isinstance(ord, Number) and ord > 0:
+            return self._array_context.np.sum(abs(ary)**ord)**(1/ord)
+        else:
+            raise NotImplementedError(f"unsupported value of 'ord': {ord}")
```

Comment:
There's nothing specific to the array context in here. This can move to the base class.

In general, please avoid copying this much code, particularly verbatim like in this case.

Reply:
Refactored in https://github.com/inducer/arraycontext/pull/14/commits/ed2374fda1e00f74bc7c91943f9da1a0da2003f5